### PR TITLE
Shallow copy resources and resource factories instead of using context chain lookup

### PR DESCRIPTION
**Pull request by agronholm**
This prevents background tasks from having access to resources that would be torn down before the tasks themselves.

### CODE REVIEW

**Comment by agronholm** [comment]
I may need to validate this against my commercial app to make sure this doesn't actually break any major design patterns we're using.

File: src/asphalt/core/_context.py [hunk 1]
```diff
@@ -277,7 +272,7 @@ async def __aenter__(self) -> Self:
                 exit_stack.callback(_current_context.reset, _reset_token)
 
                 # If this is the root task group, create and enter a task group
```
Comment (davidbrochart) [inline src/asphalt/core/_context.py:274]:
```suggestion
                # If this is the root context, create and enter a task group
```

**Review by davidbrochart — APPROVED** [review]

File: src/asphalt/core/_context.py [hunk 2]
```diff
@@ -643,21 +650,120 @@ def get_resources(self, type: type[T_Resource]) -> Mapping[str, T_Resource]:
...
+        *,
+        teardown_action: TeardownAction = "cancel",
+    ) -> Any:
+        """
+        Start a background task that gets shut down when the context shuts down.
+
+        This function is meant to be used by components to run their tasks like network
+        services that should be shut down with the application, because each call to this
+        functions registers a context teardown callback that waits for the service task to
+        finish before allowing the context teardown to continue..
```
Comment (davidbrochart) [inline src/asphalt/core/_context.py:705]:
```suggestion
        finish before allowing the context teardown to continue.
```

File: src/asphalt/core/_context.py [hunk 3]
```diff
@@ -863,6 +969,72 @@ def get_resource_nowait(
...
+    *,
+    teardown_action: TeardownAction = "cancel",
+) -> Any:
+    """
+    Start a background task that gets shut down when the context shuts down.
+
+    This function is meant to be used by components to run their tasks like network
+    services that should be shut down with the application, because each call to this
+    functions registers a context teardown callback that waits for the service task to
+    finish before allowing the context teardown to continue..
```
Comment (davidbrochart) [inline src/asphalt/core/_context.py:1014]:
```suggestion
    finish before allowing the context teardown to continue.
```

**Review by davidbrochart — APPROVED** [review]

**Pull request merged**